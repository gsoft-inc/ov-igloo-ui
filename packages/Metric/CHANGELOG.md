# @igloo-ui/metric

## 0.2.0

### Minor Changes

- 412cbc8: Added the new brand styling to Metric and VerticalBarChart.

## 0.1.3

### Patch Changes

- 08de993: Updated react-aria packages

## 0.1.2

### Patch Changes

- bf4ddef: Updated storybook version and linting. Made linting fixes in these components.
- Updated dependencies [bf4ddef]
  - @igloo-ui/tooltip@3.1.6

## 0.1.1

### Patch Changes

- c3efaf5: Fixed tests and snapshots since react-aria generated ids changed and viewBox was added to icons.

## 0.1.0

### Minor Changes

- 812961a: The initial release of the metric component

### Patch Changes

- 8faaf4c: Added missing tooltip dependency to Metric.
