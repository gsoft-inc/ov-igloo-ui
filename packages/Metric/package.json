{
  "name": "@igloo-ui/metric",
  "version": "0.4.3",
  "main": "dist/Metric.js",
  "module": "dist/Metric.js",
  "types": "dist/Metric.d.ts",
  "styles": "dist/metric.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/workleap/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Vicky Comeau vicky.comeau@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "@igloo-ui/provider": ">=0.1.5",
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/icons": "^1.12.2",
    "@igloo-ui/tokens": "^2.1.0",
    "@igloo-ui/tooltip": "^3.5.6",
    "@hopper-ui/tokens": "^4.3.2",
    "classnames": "^2.3.2",
    "react-aria": "^3.32.1"
  },
  "devDependencies": {
    "@igloo-ui/provider": "^0.1.5"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
