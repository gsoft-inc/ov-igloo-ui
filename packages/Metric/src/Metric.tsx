import * as React from "react";
import cx from "classnames";
import { useButton, type AriaButtonProps, FocusRing } from "react-aria";

import HelpSolid from "@igloo-ui/icons/dist/HelpSolid";
import Help from "@igloo-ui/icons/dist/Help";
import Alignment from "@igloo-ui/icons/dist/Alignment";
import Checkmark from "@igloo-ui/icons/dist/Checkmark";

import Tooltip from "@igloo-ui/tooltip";
import Score from "./Score";

import "./metric.scss";

export type Appearance = "positive" | "negative" | "selected";
export type MetricType = "score" | "fluctuate" | "subMetric";

export interface MetricProps extends Omit<AriaButtonProps, "type"> {
    /** The appearance of the metric */
    appearance?: Appearance;
    /** Add a specific class to the metric */
    className?: string;
    /** Add a data-test tag for automated tests */
    dataTest?: string;
    /** The icon to display */
    icon?: React.ReactElement;
    /** The label or metric name to display */
    label: string;
    /** The tooltip to display */
    tooltip?: string;
    /** The type of metric */
    type?: MetricType;
    /** The value of the metric (Only variation is used for type fluctuate) */
    value?: number | null;
    /** The variation of the metric */
    variation?: number;
}

const Metric: React.FunctionComponent<MetricProps> = ({
    appearance,
    className,
    dataTest,
    icon,
    label,
    tooltip,
    type = "score",
    value,
    variation,
    ...rest
}: MetricProps) => {
    const btnRef = React.useRef<HTMLButtonElement>(null);

    const getBrand = (): string => {
        return document.documentElement.getAttribute("data-brand") ?? "igloo";
    };

    const isWorkleap = getBrand() === "workleap";

    const { buttonProps } = useButton(rest, btnRef);

    const classes = cx("ids-metric", className, {
        [`ids-metric--${appearance}`]: appearance,
        [`ids-metric--${type}`]: type
    });
    const visualClasses = cx("ids-metric__visual", {
        [`ids-metric__visual--${appearance}`]: appearance
    });

    return (
        <FocusRing focusRingClass="keyboard-focus">
            <button
                {...buttonProps}
                className={classes}
                data-test={dataTest}
                ref={btnRef}
            >
                {type === "score" && icon && (
                    <div className={visualClasses}>{icon}</div>
                )}
                <span className="ids-metric__content">
                    <span className="ids-metric__score-group">
                        {type !== "fluctuate" && (
                            <Score value={value} forceDecimal className="ids-metric__score" />
                        )}{" "}
                        <Score
                            value={variation}
                            forceDecimal={type === "fluctuate"}
                            isVariation
                            arrowSize={type === "fluctuate" ? "medium" : "small"}
                            isSelected={appearance === "selected"}
                            hideIfZero={type !== "fluctuate"}
                            className="ids-metric__score"
                        />
                    </span>
                    <span className="ids-metric__name">
                        {type === "fluctuate" && (
                            <Alignment size="small" className="ids-metric__alignment-icon" />
                        )}
                        {label}
                    </span>
                </span>
                {(tooltip || type !== "subMetric") && (
                    <span
                        className={cx("ids-metric__status", {
                            "ids-metric__status--tooltip":
                tooltip && appearance !== "selected"
                        })}
                    >
                        {tooltip && appearance !== "selected" && (
                            <Tooltip content={tooltip}>
                                {isWorkleap ? (
                                    <Help size="small" />) : <HelpSolid size="small" />}
                            </Tooltip>
                        )}
                        {type !== "subMetric" && appearance === "selected" && (
                            <Checkmark size="small" className="ids-metric__checkmark" />
                        )}
                    </span>
                )}
            </button>
        </FocusRing>
    );
};

export default Metric;
