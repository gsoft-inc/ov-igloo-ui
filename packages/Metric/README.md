# Metric

A metric component is a button that visually represents a specific metric or measurement. It provides a concise and interactive way for users to access important information or track key performance indicators.

<Example is="custom" />

<ReferenceLinks is="custom" />

## Installation

To install `@igloo-ui/metric` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/metric
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/metric
```

## Usage

Then to use the component in your code just import it!

```jsx
import Metric from '@igloo-ui/metric';
import Wellness from '@igloo-ui/icons/dist/Wellness';

const [selected, setSelected] = React.useState(false);
const handleOnPress = () => {
  setSelected(!selected);
};

<Metric
  value={20}
  variation={3}
  label="Metric Name"
  icon={<Wellness size="medium" />}
  onPress={handleOnPress}
  appearance={selected ? 'selected' : 'positive'}
/>;
```
