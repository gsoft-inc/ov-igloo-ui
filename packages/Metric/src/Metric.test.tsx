/**
  * @jest-environment jsdom
  */
 import React from 'react';
 import { fireEvent, render, screen } from '@testing-library/react';
 import MockTooltip from '@igloo-ui/tooltip/src/__mocks__/Tooltip.mock';

 import Metric, {MetricProps} from './Metric';

 jest.mock('@igloo-ui/tooltip', () => ({
  __esModule: true,
  default: jest.fn(MockTooltip),
}));

const mockPropsEmpty: MetricProps = {
  className: 'test-class',
  label: 'Test Metric',
  tooltip: 'Tooltip text',
  type: 'score',
};

const mockPropsEmptyFluctuate: MetricProps = {
  className: 'test-class',
  label: 'Test Metric',
  tooltip: 'Tooltip text',
  type: 'fluctuate',
};

 const mockPropsPositive: MetricProps = {
  appearance: 'positive',
  className: 'test-class',
  icon: <div>Icon</div>,
  label: 'Test Metric',
  tooltip: 'Tooltip text',
  type: 'score',
  value: 10,
  variation: 1,
};

const mockPropsNegative: MetricProps = {
  appearance: 'negative',
  className: 'test-class',
  icon: <div>Icon</div>,
  label: 'Test Metric',
  tooltip: 'Tooltip text',
  type: 'score',
  value: 10,
  variation: -6,
};

const mockPropsSelected: MetricProps = {
  appearance: 'selected',
  className: 'test-class',
  icon: <div>Icon</div>,
  label: 'Test Metric',
  tooltip: 'Tooltip text',
  type: 'score',
  value: 10,
  variation: -5,
};

const mockPropsSubMetric: MetricProps = {
  className: 'test-class',
  label: 'Test Metric',
  tooltip: 'Tooltip text',
  type: 'subMetric',
  value: 10,
  variation: 5,
};

 const setup = (props: MetricProps = mockPropsPositive) => {
  return render(
    <Metric dataTest="ids-metric" {...props} />
  );
};

 describe('Metric', () => {
   test('should render without errors', () => {
    setup();
     const wrapper = screen.getByTestId('ids-metric');
     expect(wrapper).toBeInTheDocument();
     expect(wrapper).toHaveClass('ids-metric');
   });

   test('should render a snapshot', () => {
     const {asFragment} = setup();
     expect(asFragment()).toMatchSnapshot();
   });

   test('should render the label', () => {
    setup();
    
    const labelElement = screen.getByText('Test Metric');
    expect(labelElement).toBeInTheDocument();
  });

  test('should render the custom label', () => {
    setup({ ...mockPropsPositive, label: 'Custom Label' });
    
    const labelElement = screen.getByText('Custom Label');
    expect(labelElement).toBeInTheDocument();
  });

  test('should render the tooltip', () => {
    const {container} = setup();
    
    const trigger = container.querySelector('.ids-metric__status--tooltip .ids-tooltip__container');
    expect(trigger).toBeInTheDocument();
    if (trigger) {
      fireEvent.click(trigger);
    }
    const tooltipElement = screen.getByText('Tooltip text');
    expect(tooltipElement).toBeInTheDocument();
  });

  /* Score */

  test('should render the value', () => {
    setup();
    
    const valueElement = screen.getByText('10.0');
    expect(valueElement).toBeInTheDocument();
    expect(valueElement).toHaveClass('ids-metric__score');
  });

  /* Variation */

  test('should render the variation', () => {
    setup();
    
    const variationElement = screen.getByText('1');
    const postfix = variationElement.querySelector('.ids-score__postfix');
    expect(postfix).toHaveTextContent('pt');
    expect(variationElement).toBeInTheDocument();
    expect(variationElement).toHaveClass('ids-score--variation');
    expect(variationElement).toContainElement(variationElement.querySelector('.ids-score__arrow--positive'));
  });

  test('should render a negative variation', () => {
    setup(mockPropsNegative);
    
    const variationElement = screen.getByText('6');
    expect(variationElement).toBeInTheDocument();
    expect(variationElement).toHaveClass('ids-score--variation');
    expect(variationElement).toHaveClass('ids-score--negative');
    expect(variationElement).toContainElement(variationElement.querySelector('.ids-score__arrow--negative'));
  });

  /* Empty */

  test('should not render a variation when empty', () => {
    const { container } = setup(mockPropsEmpty);
    
    const variationElement = screen.getByText('0.0');
    expect(variationElement).toBeInTheDocument();
    const variationContainer = container.querySelector('.ids-score--variation');
    expect(variationContainer).not.toBeInTheDocument();
  });
  
  test('when type fluctuate, it should render the variation with a decimal and pts', () => {
    setup(mockPropsEmptyFluctuate);
    
    const variationElement = screen.getByText('0.0');
    expect(variationElement).toBeInTheDocument();
    expect(variationElement).toHaveClass('ids-score--variation');
    const arrow = variationElement.querySelector('.ids-score__arrow');
    expect(arrow).not.toBeInTheDocument();
    const postfix = screen.getByText('pts');
    expect(postfix).toBeInTheDocument();
  });

  /* Selected */

  test('should render the selected appearance', () => {
    setup(mockPropsSelected);
    
    const wrapper = screen.getByTestId('ids-metric');
    expect(wrapper).toHaveClass('ids-metric--selected');
    const status = wrapper.querySelector('.ids-metric__status');
    expect(status).toBeInTheDocument();
    expect(status).not.toHaveClass('ids-metric__status--tooltip');
    expect(status).toContainElement(wrapper.querySelector('.ids-metric__checkmark'));
  });

  /* Sub Metric */

  test('should render the subMetric type', () => {
    setup(mockPropsSubMetric);
    
    const wrapper = screen.getByTestId('ids-metric');
    expect(wrapper).toHaveClass('ids-metric--subMetric');
  });

});
