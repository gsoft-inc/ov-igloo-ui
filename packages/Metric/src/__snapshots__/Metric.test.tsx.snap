// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metric should render a snapshot 1`] = `
<DocumentFragment>
  <button
    class="ids-metric test-class ids-metric--positive ids-metric--score"
    data-test="ids-metric"
    type="button"
  >
    <div
      class="ids-metric__visual ids-metric__visual--positive"
    >
      <div>
        Icon
      </div>
    </div>
    <span
      class="ids-metric__content"
    >
      <span
        class="ids-metric__score-group"
      >
        <span
          class="ids-score ids-metric__score"
        >
          10.0
        </span>
         
        <span
          class="ids-score ids-metric__score ids-score--variation"
        >
          <svg
            class="ids-icon ids-score__arrow ids-score__arrow--positive"
            fill="none"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 12V5.333M5 7l2.764-2.764c.13-.13.342-.13.472 0L11 7"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            />
          </svg>
          1
          <span
            class="ids-score__postfix"
          >
             pt
          </span>
        </span>
      </span>
      <span
        class="ids-metric__name"
      >
        Test Metric
      </span>
    </span>
    <span
      class="ids-metric__status ids-metric__status--tooltip"
    >
      <span
        class="ids-tooltip__container"
        role="button"
        tabindex="0"
      >
        <svg
          class="ids-icon"
          fill="none"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M14 8A6 6 0 1 1 2 8a6 6 0 0 1 12 0ZM8 5.5a.75.75 0 0 0-.75.75.75.75 0 0 1-1.5 0 2.25 2.25 0 1 1 3 2.122V9a.75.75 0 0 1-1.5 0v-.75c0-.726.578-1.146.996-1.29A.75.75 0 0 0 8 5.5Zm0 6.25a.75.75 0 1 0 0-1.5.75.75 0 0 0 0 1.5Z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </span>
  </button>
</DocumentFragment>
`;
